File: data_structure/array/src/ArrayDemo.py
# Python
# データ構造: 配列 (Array)

class ArrayData:
    def __init__(self):
        self._data = []
    
    def get(self):
        # 要素を取得
        return self._data

    def get_index(self, item):
        # 配列内で指定された値を検索し、最初に見つかったインデックスを返す
        try:
            index = self._data.index(item)
            return index
        except ValueError:
            print(f"ERROR: {item} は範囲外です")
            return -1

    def get_value(self, index):
        # 指定されたインデックスの要素を取得する
        if 0 <= index < len(self._data):
            return self._data[index]
        else:
            print(f"ERROR: {index} は範囲外です")
            return None

    def add(self, item):
        # 配列の末尾に要素を追加する
        self._data.append(item)
        return True
    
    def remove(self, index):
        # 指定されたインデックスの要素を削除する
        if 0 <= index < len(self._data):
            self._data.pop(index)
            return True
        else:
            print(f"ERROR: {index} は範囲外です")
            return False
    
    def update(self, index, new_value):
        # 指定されたインデックスの要素を新しい値に更新する
        if 0 <= index < len(self._data):
            self._data[index] = new_value
            return True
        else:
            print(f"ERROR: {index} は範囲外です")
            return False

    def reverse(self):
        # 配列の要素を逆順にする
        self._data.reverse()
        return self._data
    
    def sort(self, descending=False):
        # 配列の要素をソートする
        self._data.sort(reverse=descending)
        return self._data
        
    def is_empty(self):
        # 配列が空かどうか
        return len(self._data) == 0
    
    def size(self):
        # 配列のサイズ（要素数）を返す
        return len(self._data)
    
    def clear(self):
        # 配列の全要素を削除する
        self._data = []
        return True

def main():
    print("Array TEST -----> start")

    print("\nnew")
    array_data = ArrayData()
    print(f"  現在のデータ: {array_data.get()}")

    print("\nadd")
    input = [10, 20, 30, 10, 40]
    for item in input:
        print(f"  入力値: {item}")
        output = array_data.add(item)
        print(f"  出力値: {output}")
        print(f"  現在のデータ: {array_data.get()}")

    print("\nsize")
    output = array_data.size()
    print(f"  出力値: {output}")

    print("\nis_empty")
    output = array_data.is_empty()
    print(f"  出力値: {output}")

    print("\nget_value")
    input = 2
    print(f"  入力値: {input}")
    output = array_data.get_value(input)
    print(f"  出力値: {output}")

    print("\nget_value")
    input = 10
    print(f"  入力値: {input}")
    output = array_data.get_value(input)
    print(f"  出力値: {output}")

    print("\nupdate")
    input = (1, 25)
    print(f"  入力値: {input}")
    output = array_data.update(*input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {array_data.get()}")

    print("\nupdate")
    input = (15, 25)
    print(f"  入力値: {input}")
    output = array_data.update(*input)
    print(f"  出力値: {output}") 
    print(f"  現在のデータ: {array_data.get()}") 

    print("\nget_index")
    input = 10
    print(f"  入力値: {input}")
    output = array_data.get_index(input)
    print(f"  出力値: {output}")

    print("\nget_index")
    input = 99
    print(f"  入力値: {input}")
    output = array_data.get_index(input)
    print(f"  出力値: {output}")

    print("\nremove")
    input = 3
    print(f"  入力値: {input}")
    output = array_data.remove(input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {array_data.get()}")

    print("\nremove")
    input = 8
    print(f"  入力値: {input}")
    output = array_data.remove(input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {array_data.get()}")

    print("\nreverse")
    output = array_data.reverse()
    print(f"  出力値: {output}")

    print("\nsort")
    print("  入力値: descending=False")
    output = array_data.sort(descending=False)
    print(f"  出力値: {output}")

    print("\nsort")
    print("  入力値: descending=True")
    output = array_data.sort(descending=True)
    print(f"  出力値: {output}")

    print("\nclear")
    output = array_data.clear()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {array_data.get()}")

    print("\nis_empty")
    output = array_data.is_empty()
    print(f"  出力値: {output}")

    print("\nArray TEST <----- end")

if __name__ == "__main__":
    main()

File: data_structure/array/src/ArrayDemo.go
// Go
// データ構造: 配列 (Array)

package main

import (
	"fmt"
	"sort"
)

// ArrayData は配列操作のためのデータ構造
type ArrayData struct {
	data []interface{}
}

// Get は要素を取得する
func (a *ArrayData) Get() []interface{} {
	return a.data
}

// GetIndex は配列内で指定された値を検索し、最初に見つかったインデックスを返す
func (a *ArrayData) GetIndex(item interface{}) int {
	for i, v := range a.data {
		if v == item {
			return i
		}
	}
	fmt.Printf("ERROR: %v は範囲外です\n", item)
	return -1
}

// GetValue は指定されたインデックスの要素を取得する
func (a *ArrayData) GetValue(index int) interface{} {
	if 0 <= index && index < len(a.data) {
		return a.data[index]
	}
	fmt.Printf("ERROR: %d は範囲外です\n", index)
	return nil
}

// Add は配列の末尾に要素を追加する
func (a *ArrayData) Add(item interface{}) bool {
	a.data = append(a.data, item)
	return true
}

// Remove は指定されたインデックスの要素を削除する
func (a *ArrayData) Remove(index int) bool {
	if 0 <= index && index < len(a.data) {
		a.data = append(a.data[:index], a.data[index+1:]...)
		return true
	}
	fmt.Printf("ERROR: %d は範囲外です\n", index)
	return false
}

// Update は指定されたインデックスの要素を新しい値に更新する
func (a *ArrayData) Update(index int, newValue interface{}) bool {
	if 0 <= index && index < len(a.data) {
		a.data[index] = newValue
		return true
	}
	fmt.Printf("ERROR: %d は範囲外です\n", index)
	return false
}

// Reverse は配列の要素を逆順にする
func (a *ArrayData) Reverse() []interface{} {
	for i, j := 0, len(a.data)-1; i < j; i, j = i+1, j-1 {
		a.data[i], a.data[j] = a.data[j], a.data[i]
	}
	return a.data
}

// Sort は配列の要素をソートする
func (a *ArrayData) Sort(descending bool) []interface{} {
	// インターフェース型のスライスを整数型のスライスに変換
	intData := make([]int, 0, len(a.data))
	for _, v := range a.data {
		if intVal, ok := v.(int); ok {
			intData = append(intData, intVal)
		}
	}

	if descending {
		sort.Sort(sort.Reverse(sort.IntSlice(intData)))
	} else {
		sort.Ints(intData)
	}

	// 整数型のスライスをインターフェース型のスライスに戻す
	for i, v := range intData {
		a.data[i] = v
	}

	return a.data
}

// IsEmpty は配列が空かどうかを確認する
func (a *ArrayData) IsEmpty() bool {
	return len(a.data) == 0
}

// Size は配列のサイズ（要素数）を返す
func (a *ArrayData) Size() int {
	return len(a.data)
}

// Clear は配列の全要素を削除する
func (a *ArrayData) Clear() bool {
	a.data = []interface{}{}
	return true
}

func main() {
	fmt.Println("Array TEST -----> start")

	fmt.Println("\nnew")
	arrayData := &ArrayData{}
	fmt.Printf("  現在のデータ: %v\n", arrayData.Get())

	fmt.Println("\nadd")
	input := []int{10, 20, 30, 10, 40}
	for _, item := range input {
		fmt.Printf("  入力値: %d\n", item)
		output := arrayData.Add(item)
		fmt.Printf("  出力値: %v\n", output)
		fmt.Printf("  現在のデータ: %v\n", arrayData.Get())
	}

	fmt.Println("\nsize")
	output := arrayData.Size()
	fmt.Printf("  出力値: %d\n", output)

	fmt.Println("\nis_empty")
	isEmpty := arrayData.IsEmpty()
	fmt.Printf("  出力値: %v\n", isEmpty)

	fmt.Println("\nget_value")
	indexInput := 2
	fmt.Printf("  入力値: %d\n", indexInput)
	valueOutput := arrayData.GetValue(indexInput)
	fmt.Printf("  出力値: %v\n", valueOutput)

	fmt.Println("\nget_value")
	outOfRangeIndex := 10
	fmt.Printf("  入力値: %d\n", outOfRangeIndex)
	outOfRangeOutput := arrayData.GetValue(outOfRangeIndex)
	fmt.Printf("  出力値: %v\n", outOfRangeOutput)

	fmt.Println("\nupdate")
	updateIndex, updateValue := 1, 25
	fmt.Printf("  入力値: (%d, %d)\n", updateIndex, updateValue)
	updateOutput := arrayData.Update(updateIndex, updateValue)
	fmt.Printf("  出力値: %v\n", updateOutput)
	fmt.Printf("  現在のデータ: %v\n", arrayData.Get())

	fmt.Println("\nupdate")
	invalidIndex, updateValue := 15, 25
	fmt.Printf("  入力値: (%d, %d)\n", invalidIndex, updateValue)
	invalidUpdateOutput := arrayData.Update(invalidIndex, updateValue)
	fmt.Printf("  出力値: %v\n", invalidUpdateOutput)
	fmt.Printf("  現在のデータ: %v\n", arrayData.Get())

	fmt.Println("\nget_index")
	findItem := 10
	fmt.Printf("  入力値: %d\n", findItem)
	findOutput := arrayData.GetIndex(findItem)
	fmt.Printf("  出力値: %d\n", findOutput)

	fmt.Println("\nget_index")
	notFoundItem := 99
	fmt.Printf("  入力値: %d\n", notFoundItem)
	notFoundOutput := arrayData.GetIndex(notFoundItem)
	fmt.Printf("  出力値: %d\n", notFoundOutput)

	fmt.Println("\nremove")
	removeIndex := 3
	fmt.Printf("  入力値: %d\n", removeIndex)
	removeOutput := arrayData.Remove(removeIndex)
	fmt.Printf("  出力値: %v\n", removeOutput)
	fmt.Printf("  現在のデータ: %v\n", arrayData.Get())

	fmt.Println("\nremove")
	invalidRemoveIndex := 8
	fmt.Printf("  入力値: %d\n", invalidRemoveIndex)
	invalidRemoveOutput := arrayData.Remove(invalidRemoveIndex)
	fmt.Printf("  出力値: %v\n", invalidRemoveOutput)
	fmt.Printf("  現在のデータ: %v\n", arrayData.Get())

	fmt.Println("\nreverse")
	reverseOutput := arrayData.Reverse()
	fmt.Printf("  出力値: %v\n", reverseOutput)

	fmt.Println("\nsort")
	fmt.Println("  入力値: descending=false")
	sortAscOutput := arrayData.Sort(false)
	fmt.Printf("  出力値: %v\n", sortAscOutput)

	fmt.Println("\nsort")
	fmt.Println("  入力値: descending=true")
	sortDescOutput := arrayData.Sort(true)
	fmt.Printf("  出力値: %v\n", sortDescOutput)

	fmt.Println("\nclear")
	clearOutput := arrayData.Clear()
	fmt.Printf("  出力値: %v\n", clearOutput)
	fmt.Printf("  現在のデータ: %v\n", arrayData.Get())

	fmt.Println("\nis_empty")
	isEmptyAfterClear := arrayData.IsEmpty()
	fmt.Printf("  出力値: %v\n", isEmptyAfterClear)

	fmt.Println("\nArray TEST <----- end")
}